### ISSUE TITLE
invalid version when migrating from cassandra 2.1.11

### ISSUE THREAD

**Issue by locked** [scylladb/scylla#529]
When trying to start a scylla node with the exact same data as cassandra 2.1.11, I am getting this (fatal) error:

Exiting on unhandled exception of type 'sstables::malformed_sstable_exception': invalid version

Could you add more detail about the sstable file causing the error ? Oddly, with strace, I do not see which file was opened.

**Comment by avikivity** [comment]
Are all your sstables "ka" format?

@glommer 

You're right that we should include more information with the sstable.  For strace, you need 'strace -f'.

**Comment by glommer** [comment]
I will send a patch shortly to add more information to that exception.

**Comment by locked** [comment]
Thanks, the last file opened according to strace (and thus probably causing the issue ?) seems to be a system sstable: /var/lib/scylla/data/system/schema_columnfamilies-45f5b36024bc3f83a3631034ea4fa697/system-schema_columnfamilies-ka-275-Index.db
I can upload more debug info if need be, and will try again with the patch displaying more information.

**Comment by avikivity** [comment]
Because things happen in parallel in Scylla, the last file opened may not mean much.

The patch is now committed, so if you built from source you can just pull (remember to git submodule update --recursive) and rebuild.

Another thing to try is `find -type f | grep -v -e -ka-` to see if there are any non-ka sstables.

**Comment by locked** [comment]
Thank you, I built scylla with the patch and the files causing the issue are "manifest.json" files. It seems to be related to making snapshots (https://issues.apache.org/jira/browse/CASSANDRA-5996), and though I can remove them, I think scylla should ignore them, as we will probably not be the only ones having these files.

Now I am getting this error on startup:
~~~
WARNING: exceptional future ignored of type 'std::out_of_range': Buffer underflow
Exiting on unhandled exception of type 'std::out_of_range': Buffer underflow
~~~

Should I create a new issue so you can close this one ?

**Comment by avikivity** [comment]
Keep this one open please (for ignoring the snapshot thing).

**Comment by avikivity** [comment]
And open a new one for the error. Can you run Scylla in gdb, first setting
a breakpoint with 'catch throw std::out_of_range'?

**Comment by glommer** [comment]
We do indeed ignore manifest.json at other places, but not at startup. It makes sense to do so.

Thanks for reporting this.

**Comment by avikivity** [comment]
We should distinguish between something that is an sstable, but of an 
unsupported version, and something that is clearly not an sstable.  The 
latter can be ignore, not just manifest.json files.

Consider someone opening a file with an editor and leaving a backup behind.

**Comment by glommer** [comment]
How do you intend to do that ?

Not supported sstables include future sstables, the naming of which we don't control.

**Comment by avikivity** [comment]
We can assume the TOC will be named 
<something>-<version-name>-<generation>-TOC.txt and that non-TOC files 
will be found in the TOC.

Simply crashing because someone left a file in a directory is too fragile.

**Comment by glommer** [comment]
That's not even true for the la files, that has the generation and version names flipped.
The only thing they have in common is the presence of an extension, and "-" markers. Both facts are too fragile for a test.

We could, of course, just ignore files whose format we can't read, and print a warning about it.
The user would see that we have found a file that we don't understand - and then we can shut up the message for manifest.json specifically, since this one is expected, but not crash.

Since we won't be able to parse it, it will never be added to the list of sstables and it won't be opened. So we wouldn't be opening a file that we don't understand.

**Comment by avikivity** [comment]
Ignoring files is too dangerous, it can lead to data loss.

What does origin do?  It has the same problem.

**Comment by glommer** [comment]
I would have to look into that. The code I have here (for 2.1.8, and consequently before la) just assumes everything has the ka format, so it looks for something that matches that.

**Comment by glommer** [comment]
Here is what Cassandra 3.1 will do. The code is in fromFilename(), io/sstable/Descriptor.java

It tokenizes the file, using - as a separator. It then tries to read the tokens and based on that, discover which format it is supposed to read. *At this point*, whatever does not fit the expected tokenization throws.

This code is then called for SSTableLister in db/Directories.java, which will list files in the current directory. This will ignore any file for which it does not know the format. This is done by tryComponentFromFilename() in SSTable.java, that changes that exception to a null return which is then ignored by the SSTableLister.

**Comment by glommer** [comment]
How do I view a possible fix for this: We don't necessarily have to use a tokenizer for this. Our regexp match will work just as well. Anything that does not match any known format is then ignored where it makes sense to do so (in our equivalent to the SSTableLister)

**Comment by glommer** [comment]
Main disadvantage of the regexp over the tokenizer: it won't protect us against future versions (say, if somebody copies 3.0 files to us)

**Comment by tzach** [comment]
This issue  cause 'nodetool refresh' to fail if not removing manifest.json first.
